RedHatInsights/rhsm-subscriptions
# ENT-3517: add core hours hosts apis

Update host APIs for monthly core-hours

* Quick & dirty implementation for including month core-hours in the payload of GET /hosts.
* Added optional `beginning` and `ending` query parameters to the API.  These are ignored if the product_id isn't `OpenShift-dedicated-metrics` or `OpenShift-metrics`.
* When the product_id is one of these, validation is in place to make sure that `ending` doesn't come before `beginning`, and that `beginning` and `ending` are of the same month. There's a very strong assumption here that the UI is always going to pass the first timestamp of a month and the last timestamp of the same month.
* Also as a part of this PR, I introduced using JPA Specifications/Criteria to the `HostRepository` class.  This will be helpful/reused when we come back to do the hypervisor work, as well as refactoring to to use the Measurments.Uom/Value pattern in lieu of the now deprecated cores/sockets fields in the `Host` entity.  In the interest of time, I took the shortcut and introduced the `HostRepository.findAll(Specification, Pageable)` and `Host.asTallyHostViewApiHost` methods, and invoke those instead when the product_id is one of the designated types.

Testing - Prep Some Data
-------
1. Clear some data and then ensure opted in:

```sql
truncate table events;
truncate table tally_snapshots cascade;
truncate table hosts cascade;
insert into account_config(account_number, sync_enabled, reporting_enabled, opt_in_type, created, updated) VALUES ('account123', true, true, 'DB', now(), now());
```

2. Insert some events:

```sql
insert into events(id, account_number, timestamp, data) values ('033a45b7-e8c4-440a-b0bb-76c3f3f67d85','account123', '2021-01-01T00:00Z', '{"account_number":"account123","timestamp":"2021-01-01T00:00:00Z","event_id":"033a45b7-e8c4-440a-b0bb-76c3f3f67d85","service_type":"OpenShift Cluster","instance_id":"4e8534b5-42b5-4f00-9969-29046e2b1986","measurements": [{"uom":"Cores","value":4.0}]}');
insert into events(id, account_number, timestamp, data) values ('a66ac2a1-17c0-4bc3-bc29-0430953d626c','account123', '2021-01-01T02:00Z', '{"account_number":"account123","timestamp":"2021-01-01T02:00:00Z","event_id":"a66ac2a1-17c0-4bc3-bc29-0430953d626c","service_type":"OpenShift Cluster","instance_id":"4e8534b5-42b5-4f00-9969-29046e2b1986","measurements": [{"uom":"Cores","value":2.0}]}');
insert into events(id, account_number, timestamp, data) values ('fad48795-ae0a-4674-b864-822ddfb799d0','account123', '2021-01-01T03:00Z', '{"account_number":"account123","timestamp":"2021-01-01T03:00:00Z","event_id":"fad48795-ae0a-4674-b864-822ddfb799d0","service_type":"OpenShift Cluster","instance_id":"4e8534b5-42b5-4f00-9969-29046e2b1986","measurements": [{"uom":"Cores","value":4.0}]}');
```

3. Invoke the hourly tally for the account:

```
curl 'http://localhost:8080/actuator/hawtio/jolokia/' -H 'Content-Type: application/json' -d '{"type":"exec","mbean":"org.candlepin.subscriptions.jmx:name=tallyJmxBean,type=TallyJmxBean","operation":"tallyAccountByHourly(java.lang.String,java.lang.String,java.lang.String)","arguments":["account123","2021-01-01T00:00:00Z","2021-01-01T23:59:59.999Z"]}'
```

4. Verify the host records contains 10 cores:

```sql
select * from instance_monthly_totals;
```

5. Insert a new event in the middle:

```sql
insert into events(id, account_number, timestamp, data) values ('c8e49ded-cd75-4016-8be7-3f795ff65645','account123', '2021-01-01T00:00Z', '{"account_number":"account123","timestamp":"2021-01-01T01:00:00Z","event_id":"033a45b7-e8c4-440a-b0bb-76c3f3f67d85","service_type":"OpenShift Cluster","instance_id":"4e8534b5-42b5-4f00-9969-29046e2b1986","measurements": [{"uom":"Cores","value":20.0}]}');
```

6. Retrigger the range containing only the new event

```
curl 'http://localhost:8080/actuator/hawtio/jolokia/' -H 'Content-Type: application/json' -d '{"type":"exec","mbean":"org.candlepin.subscriptions.jmx:name=tallyJmxBean,type=TallyJmxBean","operation":"tallyAccountByHourly(java.lang.String,java.lang.String,java.lang.String)","arguments":["account123","2021-01-01T00:00:00Z","2021-01-01T02:00:00Z"]}'
```

7. Observe that the new monthly total contains 30 cores:

```sql
select * from instance_monthly_totals;
```

Testing - Call Hosts API
-------
1. Call the API with the OpenShift-metrics product and a valid beginning/ending date range
```bash
curl -X GET --location "http://localhost:8080/api/rhsm-subscriptions/v1/hosts/products/OpenShift-metrics?beginning=2021-01-01T00:00:00.000Z&ending=2021-01-31T23:59:59.999Z" \
    -H "x-rh-identity: eyJpZGVudGl0eSI6eyJhY2NvdW50X251bWJlciI6ImFjY291bnQxMjMiLCJ0eXBlIjoiVXNlciIsInVzZXIiOnsiaXNfb3JnX2FkbWluIjp0cnVlfSwiaW50ZXJuYWwiOnsib3JnX2lkIjoib3JnMTIzIn19fQo"
```

2. Observe the response is populated correctly with `core_hours`
```json
{"data":[{"display_name":"4e8534b5-42b5-4f00-9969-29046e2b1986","sockets":0,"cores":4,"core_hours":30.0,"hardware_type":"PHYSICAL","measurement_type":"PHYSICAL","last_seen":"2020-12-31T22:00:00-05:00","is_unmapped_guest":false,"is_hypervisor":false}],"meta":{"count":1,"product":"OpenShift-metrics"}}
```
